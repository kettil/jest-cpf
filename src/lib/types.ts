import { thresholdKeys } from './threshold/settings';

export type ExtendPromiseType<T> = { promise: Promise<T | Error>; isPending: () => boolean };

export type ArgvObjectType = { config?: string; [x: string]: unknown };

export type RunnerType = {
  configFile: string;
  coverageFile: string;

  test?: {
    file: string;
  } & ExtendPromiseType<string>;
};

export type RunnerEvaluationDataType = {
  thresholds: ThresholdGroupType[];
  config: JestConfigType;
  barIncrement: BarReturnType['increment'];
};

export type RunnerInitializationDataType = {
  testFiles: string[];
  argvArray: string[];
  configFile: string;
  config: JestConfigType;
  command: string;
};

export type RunnerDataType = RunnerEvaluationDataType & RunnerInitializationDataType;

export type BarReturnType = {
  increment: () => void;
  stop: () => void;
};

export type JestConfigType = {
  testMatch: string[];
  rootDir: string;
  roots: string[];

  collectCoverage: true;
  collectCoverageFrom: string[];
  coverageDirectory: string;
  coverageReporters: string[];
  coverageThreshold: { global: ThresholdType<number> };
  coveragePathIgnorePatterns: string[];

  [x: string]: unknown;
};

export type CoverageSummaryType = {
  pct: number;
  total: number;
  covered: number;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ThresholdKeysType = typeof thresholdKeys[any];
export type ThresholdColorType = 'success' | 'error';
export type ThresholdErrorTypes = 'ERROR';
export type ThresholdValueType = { percent: number; uncovered: number } | ThresholdErrorTypes;
export type ThresholdRowType = string[];
export type ThresholdType<T = ThresholdValueType> = Record<ThresholdKeysType, T>;
export type ThresholdGroupType<T = ThresholdValueType> = {
  file: string;
  threshold: ThresholdType<T>;
  error?: unknown;
};

//
// helper
//

export type UnpackedPromise<T> = T extends Promise<infer U> ? U : T;
