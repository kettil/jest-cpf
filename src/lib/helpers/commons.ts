import { ExtendPromiseType } from '../types';

export const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms, ms));

export const extendPromise = <T>(promise: Promise<T>): ExtendPromiseType<T | Error> => {
  let isPending = true;

  return {
    isPending: () => isPending,
    promise: new Promise<T | Error>((resolve) => {
      promise.then(
        (v: T) => {
          isPending = false;
          resolve(v);
        },
        (error: Error) => {
          isPending = false;
          resolve(error);
        },
      );
    }),
  };
};
