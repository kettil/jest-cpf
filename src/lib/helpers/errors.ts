export class TestError extends Error {
  readonly data: string;

  constructor(message?: string, data = '') {
    super(message);

    this.data = data;
  }
}
