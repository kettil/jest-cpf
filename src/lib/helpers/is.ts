import { CoverageSummaryType } from '../types';

export const isObject = (data: unknown): data is Record<string, unknown> => typeof data === 'object' && data !== null;

export const isSummary = (data: unknown): data is CoverageSummaryType =>
  isObject(data) && typeof data.pct === 'number' && typeof data.total === 'number' && typeof data.covered === 'number';

export const isGlobalThreshold = (threshold: unknown): threshold is { global: Record<string, unknown> } =>
  isObject(threshold) && isObject(threshold.global);
