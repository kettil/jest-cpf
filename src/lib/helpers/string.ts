import { join, dirname, basename } from 'path';

export const cleanFileFromRootPath = (file: string, rootDirectory: string) =>
  file.slice(Math.max(0, rootDirectory.length + 1));

export const cleanFileFromExtensions = (file: string) => {
  const path = dirname(file);
  const base = basename(file);
  const name = base.slice(0, Math.max(0, base.indexOf('.') + 1));

  return join(path, name);
};

export const mapPerLine = (value: string, callback: (line: string) => string) =>
  value
    .split('\n')
    .map(callback)
    .join('\n');

export const padFill = (value: string, length: number, fill: string) =>
  [...new Array(Math.max(0, length - value.length))].fill(fill).join('');

export const padLeft = (value: string, length: number, fill: string): string => padFill(value, length, fill) + value;

export const padRight = (value: string, length: number, fill: string): string => value + padFill(value, length, fill);

export const formatPercent = (value: number): string => {
  const [pre, post = ''] = value.toString().split('.');

  return [padLeft(pre, 3, ' '), padRight(post, 2, '0')].join('.');
};
