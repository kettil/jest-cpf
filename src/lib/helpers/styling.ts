import { ThresholdColorType } from '../types';
import chalk = require('chalk');

type ChalkType = 'underline' | 'red' | 'green' | 'bold';

export const statusToType = (status: ThresholdColorType): ChalkType => {
  switch (status) {
    case 'success':
      return 'green';

    case 'error':
    default:
      return 'red';
  }
};

const styling = (text: string | number, type?: ChalkType): string => {
  if (typeof text === 'number') {
    text = text.toString();
  }

  switch (type) {
    case 'green':
      return chalk.green(text);

    case 'red':
      return chalk.red(text);

    case 'underline':
      return chalk.underline(text);

    case 'bold':
      return chalk.bold(text);

    default:
      return text;
  }
};

export default styling;
