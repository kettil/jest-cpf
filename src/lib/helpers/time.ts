import { DateTime } from 'luxon';

export const startTime = () => DateTime.utc();

export const getRunTime = (time: DateTime) => {
  const runtime = DateTime.utc().diff(time, ['minutes', 'seconds']);

  return runtime.toFormat('mm:ss');
};
