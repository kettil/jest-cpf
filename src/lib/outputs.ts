/* eslint-disable no-console */
import { SingleBar } from 'cli-progress';
import { table } from 'table';
import { TestError } from './helpers/errors';
import { padLeft, mapPerLine } from './helpers/string';
import styling from './helpers/styling';
import createTableData from './threshold/createTableData';
import { tableConfig } from './threshold/settings';
import { ThresholdColorType, ThresholdGroupType, BarReturnType, ThresholdType } from './types';

export const outputError = (error: unknown) => {
  if (error instanceof Error) {
    if (error instanceof TestError) {
      console.error(styling('|> An error was thrown during testing!', 'red'));
      console.error(styling('|> ', 'red'));
      console.error(styling('|> Output', 'red'));
      console.error(styling('|> ', 'red'));
      console.error(mapPerLine(error.data, (line) => styling('|> ', 'red') + line));
    } else {
      console.error(mapPerLine(error.stack ?? `Error: ${error.message}`, (line) => styling('|> ', 'red') + line));
    }
  } else {
    console.error(error);
  }

  console.log('');
};

export const outputMetrics = (status: Record<ThresholdColorType, number>, runtime: string) => {
  const types = Object.keys(status) as ThresholdColorType[];
  const maxChars = types.reduce((p, c) => Math.max(p, status[c]), 0).toString().length;

  console.log(styling('Summary', 'underline'));
  console.log(`├─ Success: ${styling(padLeft(status.success.toString(), maxChars, ' '), 'green')}`);
  console.log(`└─ Error:   ${styling(padLeft(status.error.toString(), maxChars, ' '), 'red')}`);
  console.log('');

  console.log(`Runtime: ${runtime} (mm:ss)`);
  console.log('');
};

export const outputProgress = (count: number): BarReturnType => {
  const bar = new SingleBar({
    format: `${styling('{bar}', 'red')} {percentage}% | {value}/{total} Files`,
    barCompleteChar: '\u2588',
    barIncompleteChar: '\u2591',
    hideCursor: true,
  });

  bar.start(count, 0);

  return {
    increment: () => bar.increment(),
    stop: () => {
      bar.stop();
      console.log('');
    },
  };
};

export const outputStatus = (countFiles: number, countTestFiles: number, cpuCount: number, cpuUsed: number) => {
  console.log('');
  console.log(`CPU detected: ${styling(cpuCount, 'red')} (${styling(cpuUsed, 'red')} are used)`);
  console.log(`Files found:  ${styling(countFiles, 'red')} (${styling(countTestFiles, 'red')} with tests)`);
  console.log('');
};

export const outputTable = (thresholds: ThresholdGroupType[], thresholdLimits: ThresholdType<number>) => {
  const thresholdRows = createTableData(thresholds, thresholdLimits);

  console.log(table(thresholdRows, tableConfig));
};
