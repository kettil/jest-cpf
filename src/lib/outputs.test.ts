/* eslint-disable no-console, @typescript-eslint/unbound-method */
import { SingleBar } from 'cli-progress';
import { TestError } from './helpers/errors';
import styling, { statusToType } from './helpers/styling';
import { outputError, outputMetrics, outputProgress, outputStatus, outputTable } from './outputs';
import { ThresholdGroupType } from './types';

jest.mock('cli-progress');
jest.mock('./helpers/styling');
jest.spyOn(console, 'log');
jest.spyOn(console, 'error');

beforeEach(() => {
  (styling as jest.Mock).mockImplementation((m: string, t: string) => `[${t}]${m}[/${t}]`);
  (statusToType as jest.Mock).mockImplementation((s: string) => (s === 'success' ? 'green' : 'red'));
});

describe('outputError()', () => {
  test('it should work with test error object', () => {
    const error = new TestError('test-error-object', 'data-string');

    outputError(error);

    expect(console.error).toHaveBeenCalledTimes(5);
    expect(console.error).toMatchSnapshot('error');
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(console.log).toHaveBeenCalledWith('');
  });

  test('it should work with error object', () => {
    const error = new Error('error-object-with-stack');

    outputError(error);

    expect(console.error).toHaveBeenCalledTimes(1);
    expect(console.error).toHaveBeenCalledWith(expect.any(String));
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(console.log).toHaveBeenCalledWith('');
  });

  test('it should work with error object and without stack', () => {
    const error = new Error('error-object-without-stack');

    delete error.stack;

    outputError(error);

    expect(console.error).toHaveBeenCalledTimes(1);
    expect(console.error).toMatchSnapshot('error');
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(console.log).toHaveBeenCalledWith('');
  });

  test('it should work with error string', () => {
    outputError('error-string');

    expect(console.error).toHaveBeenCalledTimes(1);
    expect(console.error).toHaveBeenCalledWith('error-string');
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(console.log).toHaveBeenCalledWith('');
  });
});

describe('outputMetrics()', () => {
  test('it should work with test error object', () => {
    const status = { error: 6, success: 34 };

    outputMetrics(status, '00:00');

    expect(console.error).toHaveBeenCalledTimes(0);
    expect(console.log).toHaveBeenCalledTimes(6);
    expect(console.log).toMatchSnapshot('log');
  });
});

describe('outputProgress()', () => {
  test('it should work', () => {
    const bar = outputProgress(10);

    expect(bar).toEqual({ increment: expect.any(Function), stop: expect.any(Function) });

    expect(SingleBar.prototype.start).toHaveBeenCalledTimes(1);
  });

  test('it should work and increment the bar', () => {
    const bar = outputProgress(10);

    expect(bar.increment).toEqual(expect.any(Function));

    bar.increment();

    expect(SingleBar.prototype.increment).toHaveBeenCalledTimes(1);
  });

  test('it should work and stop the bar', () => {
    const bar = outputProgress(10);

    expect(bar.stop).toEqual(expect.any(Function));

    bar.stop();

    expect(SingleBar.prototype.stop).toHaveBeenCalledTimes(1);
    expect(console.log).toHaveBeenCalledTimes(1);
  });
});

describe('outputStatus()', () => {
  test('it should work', () => {
    outputStatus(5, 3, 4, 2);

    expect(console.error).toHaveBeenCalledTimes(0);
    expect(console.log).toHaveBeenCalledTimes(4);
    expect(console.log).toMatchSnapshot('log');
  });
});

describe('outputTable()', () => {
  test('it should work', () => {
    const thresholdLimits = { branches: 90, functions: 90, statements: 90, lines: 90 };
    const thresholds: ThresholdGroupType[] = [
      {
        file: '/path/to/file-1.ts',
        threshold: {
          branches: { percent: 90, uncovered: 3 },
          functions: { percent: 90, uncovered: 3 },
          lines: { percent: 80, uncovered: 3 },
          statements: { percent: 90, uncovered: 3 },
        },
      },
      {
        file: '/path/to/file-1.ts',
        threshold: { branches: 'ERROR', functions: 'ERROR', lines: 'ERROR', statements: 'ERROR' },
      },
    ];

    outputTable(thresholds, thresholdLimits);

    expect(console.error).toHaveBeenCalledTimes(0);
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(console.log).toMatchSnapshot('log');
  });
});
