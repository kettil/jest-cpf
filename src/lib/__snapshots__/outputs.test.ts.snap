// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outputError() it should work with error object and without stack: error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[red]|> [/red]Error: error-object-without-stack",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`outputError() it should work with test error object: error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[red]|> An error was thrown during testing![/red]",
    ],
    Array [
      "[red]|> [/red]",
    ],
    Array [
      "[red]|> Output[/red]",
    ],
    Array [
      "[red]|> [/red]",
    ],
    Array [
      "[red]|> [/red]data-string",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`outputMetrics() it should work with test error object: log 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[underline]Summary[/underline]",
    ],
    Array [
      "├─ Success: [green]34[/green]",
    ],
    Array [
      "└─ Error:   [red] 6[/red]",
    ],
    Array [
      "",
    ],
    Array [
      "Runtime: 00:00 (mm:ss)",
    ],
    Array [
      "",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`outputStatus() it should work: log 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "",
    ],
    Array [
      "CPU detected: [red]4[/red] ([red]2[/red] are used)",
    ],
    Array [
      "Files found:  [red]5[/red] ([red]3[/red] with tests)",
    ],
    Array [
      "",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`outputTable() it should work: log 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "────────────────┬─────────────────────────────────┬───────────────────────┬───────────────────────┬───────────────────────┬──────────────────────
 [bold]#[/bold] │ [bold]File[/bold]               │  [bold]% Stmts[/bold] │ [bold]% Branch[/bold] │  [bold]% Funcs[/bold] │ [bold]% Lines[/bold] 
────────────────┼─────────────────────────────────┼───────────────────────┼───────────────────────┼───────────────────────┼──────────────────────
              1 │ /path/to/[bold]file-1.ts[/bold] │ [green] 90.00[/green] │ [green] 90.00[/green] │ [green] 90.00[/green] │ [red] 80.00[/red]    
              2 │ /path/to/[bold]file-1.ts[/bold] │      [red]ERROR[/red] │      [red]ERROR[/red] │      [red]ERROR[/red] │ [red]ERROR[/red]     
────────────────┴─────────────────────────────────┴───────────────────────┴───────────────────────┴───────────────────────┴──────────────────────
",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
